### ISSUE TITLE
Simplify serialization and increase size of max install message

### ISSUE THREAD

**Issue by fcangialosi** [ccp-project/libccp#17]
A lot of the serialization code calls memcpy a number of times when it could just call it once, e.g. serialize.c:

```c
    memcpy(msg, buf, 2 * sizeof(u32));
    buf += 2 * sizeof(u32);

    memcpy(&msg->exprs, buf, msg->num_expressions * sizeof(struct ExpressionMsg));
    buf += msg->num_expressions * sizeof(struct ExpressionMsg);
    memcpy(&msg->instrs, buf, msg->num_instructions * sizeof(struct InstructionMsg));
    buf += msg->num_expressions * sizeof(struct InstructionMsg);
```

and read_msg could just cast the buffer to an installexpressionmsg rather than allocating space for one on the stack and copying it over. then we're only limited by the size of the message (which is 32K for netlink)

**Comment by deeptir18** [comment]
So there's one issue with the snippet you've put above - the way portus sends down messages is the following:

- 3 ints (uid, num instructions, num expressions)
Then the array of expressions, and immediately, without padding, the number of instructions.
You can't do a simple memcpy because InstallExpressionMsg assumes there's padding (there could be MAX_EXPRESSIONS number of expressions), which is why we have to do multiple memcpys and move the pointer.

Anyway, in the new code I wrote, we just read directly from the buf. Look at: #19

**Comment by fcangialosi** [comment]
Resolved by #19
